### ISSUE TITLE
At least one of these fields is required

### ISSUE THREAD

**Issue by yanfali** [rs/rest-layer#52]
I'm trying to figure out a way to express a schema where I have a set of fields and at least one of them must be present to validate the JSON. AnyOf is for a specific Field. I'm thinking I need to either:

- modify schema to support - at least one of
- look at the map we return and check it's length - if it's zero then invariant is false

Am I missing something? Thanks

**Comment by smyrman** [comment]
@yanfali, I don't se any other way than what you mention - do a AnyOf in a subfield, which is not exactly what you want.

>  look at the map we return and check it's length - if it's zero then invariant is false

JSON Schema has a field "minProperties" that can be used for this purpose http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.14. There is also a "maxProperties".

@rs, @yanfali - would it make sense to add `MinLen`/`MaxLen` to [Dict](https://godoc.org/github.com/rs/rest-layer/schema#Dict) and [Schema](https://godoc.org/github.com/rs/rest-layer/schema#Schema)? @yanfali, I expect this would satisfy your use-case?

As for MaxLen, interpret 0 as no limit, as is already done for [String](https://godoc.org/github.com/rs/rest-layer/schema#String) and [Password](https://godoc.org/github.com/rs/rest-layer/schema#Password). The `jsonschema` package should of course use "minProperties" / "maxProperties" even if the terms MinLen/MaxLen is used within the `schema` package.

> modify schema to support - at least one of
If we where to support OneOf / AnyOf at the top level, what would make the most sense to me at least, would be if you could pass the AnyOf object directly to `resource.Bind()`, but that sounds like a breaking change:-( If we where to add AnyOf/OneOf to the Schema struct, that sounds like it would make validation, not to mention json-schema serialization, more painful. @rs, what is your thoughts?

**Comment by smyrman** [comment]
For completeness, you could add `MinLen` / `MaxLen` to [Array](https://godoc.org/github.com/rs/rest-layer/schema#Array) as well. The JSON schema terms for an array would be "minItem" / "maxItems" http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.10.

**Comment by smyrman** [comment]
@yanfali, did you have any time to consider whether adding MinLen/MaxLen to Schema (and similar types such as Dict and Array) will solve your use-case?

**Comment by smyrman** [comment]
I did a very quick patch in https://github.com/smyrman/rest-layer/tree/min-max-len you can try out if you want.

I don't know if adding these fields are the right thing to do or not - especially for Schema which is already quite complex. Object might be better, although that excludes min/max length at the top level...

**Comment by yanfali** [comment]
Sorry @smyrman been really busy with work. I'll get to this tomorrow PST
most likely. Thanks

**Comment by smyrman** [comment]
@yanfali, no worries.

**Comment by yanfali** [comment]
It looks like a decent addition, with some tests. MinProperties might work for me. I sort of have a schema which requires at least 3 parameters; 1 is required and 2 optionals, but the optionals come from different sets. Long term I'm planning on breaking these into 2 APIs.

**Comment by smyrman** [comment]
> I sort of have a schema which requires at least 3 parameters; 1 is required and 2 optionals
@yanfali, do I understand you correctly if you mean that you have:
1.  A schema definition with 1 required and N optional fields
2. When POSTing a new entry, the resulting document must contain the required fields plus any two of the optional fields.
3. When PATCHing an existing entry (if allowed), it's enough that the final (patched) document contain the required fields plus any two of the optional fields.
4. Autoset fields (if applicable) are included in the count of three, or are constant count, so the count can be added to the MinLen value.

If any of these points are not true, I don't think my patch will help...

@rs, before I go on to add tests and do a PR, do you agree with the changes propsed in https://github.com/smyrman/rest-layer/commit/8c53e1084be51585f1439d89cffcdd9924e9e808?

Basically it adds MinLen/MaxLen to Schema, Dict and Array. The part I am most unsure about is actually adding MinLen/MaxLen to Schema rather than Object, but that is necessary to support these parameters on a document root level.

**Comment by yanfali** [comment]
So right now we're using POST with `X-HTTP-Method-Override: GET` pattern to support sending a JSON body instead of a GET because browser don't support GET with bodies over XHR. At this time there's no POST or PATCH path into this API, it's read-only.

**Comment by rs** [comment]
@smyrman yes I'm fine with that.

**Comment by smyrman** [comment]
> At this time there's no POST or PATCH path into this API, it's read-only.

@yanfali, OK, so do I then understand you correctly that you don't have any immediate needs for the validation, but it might be useful later on?

Anyway I will put up a PR with tests of min/max len when I get time, as we agreed it's a decent addition.

**Comment by smyrman** [comment]
#59 adds MinLen/MaxLen to `schema.Array` only. When we agree with the general approach, I will do the same for `Dict` and `Schema`.

**Comment by smyrman** [comment]
@yanfali, with #60 merged as well, I guess we can close this issue?
